src/t_hash.c

```diff
@@ -29,52 +29,206 @@
 
 #include "server.h"
 #include <math.h>
+#include <stdbool.h>
 
 /*-----------------------------------------------------------------------------
  * Hash Entry API
  *----------------------------------------------------------------------------*/
 
-struct hashTypeEntry {
+/* The hashTypeEntry pointer is the field sds. We encode the entry layout type
+ * in the field SDS header. Field type SDS_TYPE_5 doesn't have any spare bits to
+ * encode this so we use it only for the first layout type.
+ *
+ * ENTRY_ENC_EMB_VALUE, used when it fits in a cache line. The value is stored
+ * as SDS_TYPE_8. The field can use any SDS type.
+ *
+ *     +--------------+---------------+
+ *     | field        | value         |
+ *     | hdr "foo" \0 | hdr8 "bar" \0 |
+ *     +------^-------+---------------+
+ *            |
+ *            |
+ *          entry pointer = field sds
+ *
+ * ENTRY_ENC_PTR_VALUE, with value pointer, used for larger fields and values.
+ * The field is SDS type 8 or higher.
+ *
+ *     +-------+--------------+
+ *     | value | field        |
+ *     | ptr   | hdr "foo" \0 |
+ *     +-------+------^-------+
+ *                    |
+ *                    |
+ *                 entry pointer = field sds
+ */
+
+typedef enum {
+    ENTRY_ENC_EMB_VALUE = 0,
+    ENTRY_ENC_PTR_VALUE = 1
+} hashTypeEntryEnc;
+
+#define FIELD_SDS_AUX_BIT_ENTRY_HAS_PTR_VALUE 0
+
+/* Struct used for ENTRY_ENC_PTR_VALUE. */
+typedef struct {
     sds value;
-    char field_offset;
     char field_data[];
-};
+} hashTypeEntryPtrValue;
```

Comment:
IMO this only leads to some confusion when for fully embedded entry we use some unstructured array of characters and for partially embedded we use some structured structure. suggest to remove this.

Reply:
Sure, working with char-ptr is fine when it's only one pointer and an embedded field. I can change to that.

If we'll have more fields in a future variant though, like ttl, value-ptr and embedded field, then at some point a struct is probably worth it... But that's a future problem.